src/Module.zig

```diff
@@ -92,6 +92,17 @@ embed_table: std.StringHashMapUnmanaged(*EmbedFile) = .{},
 /// is not yet implemented.
 intern_pool: InternPool = .{},
 
+/// The index type for this array is `CaptureScope.Index` and the elements here are
+/// the indexes of the parent capture scopes.
+/// Memory is owned by gpa; garbage collected.
+capture_scope_parents: std.ArrayListUnmanaged(CaptureScope.Index) = .{},
+/// Value is index of type
+/// Memory is owned by gpa; garbage collected.
+runtime_capture_scopes: std.ArrayHashMapUnmanaged(CaptureScope.Key, InternPool.Index) = .{},
+/// Value is index of value
+/// Memory is owned by gpa; garbage collected.
+comptime_capture_scopes: std.ArrayHashMapUnmanaged(CaptureScope.Index, InternPool.Index) = .{},
```

Comment:
Garbage collecting these structures is potentially expensive. For the maps, you will need to iterate keys to find all of the ones with unreferenced capture scopes, and delete all of those elements. Then when you remove the unreferenced elements from the parents array, the remaining elements may need to change value. (For example, if you have Index values 0-5 in use, and 3 is gc'd, you need to replace all references to 5 with 3 after removeSwapping the parents array). This will then require rebuilding the hash index for the hash structures for any keys which were renumbered.  Using a slot map lookup to keep keys stable would help but is potentially a lot of extra memory compared to what this saves.

Other than that, I think these changes go in a good direction.

Reply:
A multimap might be a more appropriate structure here, basically a map but if you insert multiple items with the same key it treats it like a collision and puts the items next to each other. Iterating or removing values associated with a key is trivial because they are all next to each other.

Reply:
I wasn't planning on implementing deletion for anything - rather, scanning all the roots and adding those to a new world of objects. Maybe this is a rookie GC101 mistake; I've never written a GC before.

Note that the *main* purpose of this change is to have the same in-memory representation as on-disk representation for fast and simple serialization/deserialization.

Reply:
Makes sense, seems like a good change then!